File: ruby/space-age/space_age.rb
class SpaceAge
  attr_accessor :seconds

  def initialize(seconds)
    @seconds = seconds
  end

  def on_earth
    seconds / EARTH_YEAR
  end

  private

  def method_missing(planet)
    on_earth / PLANETS[planet]
  end

  EARTH_YEAR = 31_557_600.0

  PLANETS = {
    on_mercury: 0.2408467,
    on_venus:   0.61519726,
    on_mars:    1.8808158,
    on_jupiter: 11.862615,
    on_saturn:  29.447498,
    on_uranus:  84.016846,
    on_neptune: 164.79132
  }.freeze
end

File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :uranus | :neptune

  @earth_year 31557600

  @planets %{
    mercury: 0.2408467,
    venus:   0.61519726,
    mars:    1.8808158,
    jupiter: 11.862615,
    saturn:  29.447498,
    uranus:  84.016846,
    neptune: 164.79132,
  }

  def age_on(:earth, seconds) do
    seconds / @earth_year
  end

  def age_on(planet, seconds) do
    age_on(:earth, seconds) / @planets[planet]
  end
end
